### ISSUE TITLE
Incorrect adjoint gradients with overlapping design variables

### ISSUE THREAD

**Issue by mochen4** [NanoComp/meep#1984]
When the material grids of design variables are overlapped, e.g. [the eigenmode tutorial example](https://github.com/NanoComp/meep/blob/master/python/examples/adjoint_optimization/01-Introduction.ipynb), the adjoint gradients don't agree with the finite difference gradients. 

In the aforementioned example, the gradients are correct if only one block of the design region is kept in the`` geometry`` but not both:
```
geometry=[
...
mp.Block(center=design_region.center, size=design_region.size, material=design_variables), # design region
mp.Block(center=design_region.center, size=design_region.size, material=design_variables,
             e1=mp.Vector3(x=-1).rotate(mp.Vector3(z=1), np.pi/2), e2=mp.Vector3(y=1).rotate(mp.Vector3(z=1), np.pi/2))
]
```

**Comment by mochen4** [comment]
> IIRC, the error is a constant error, right? meaning we are just off by a scale factor (i.e. it's not as if the entire gradient is just wrong)?

Yes, I think the gradients are off by a scale factor, but that scale varies a bit with different random structures.

**Comment by smartalecH** [comment]
I'm pretty sure the issue is due to the following snippet introduced in #1780.

https://github.com/NanoComp/meep/blob/11545a1b859222df2c404370673115666b589c22/src/meepgeom.cpp#L2758-L2771

**Comment by MarkMa1990** [comment]
> I'm pretty sure the issue is due to the following snippet introduced in #1780.
> 
> https://github.com/NanoComp/meep/blob/11545a1b859222df2c404370673115666b589c22/src/meepgeom.cpp#L2758-L2771

Yes, exactly. In fact, I just compared different versions, it seems that from version 1.21.0, the material derivative to the design variable u is changed: 

![image](https://user-images.githubusercontent.com/10673026/169806823-297068d1-6022-4da2-b3fc-1d5ad2d5ff1a.png)
![image](https://user-images.githubusercontent.com/10673026/169806997-de14c893-0fa7-435d-93b5-9b7b803c5ba9.png)
![image](https://user-images.githubusercontent.com/10673026/169807228-06cb63e6-6a14-47bc-a970-ce464b151c7c.png)
